lua_modules/liquidcrystal/lc-i2c4bit.lua

```diff
@@ -66,36 +63,45 @@ return function(bus_args)
                      bit.lshift(bit.isset(hi, d7) and 1 or 0, 7))
    end
 
-   -- init sequence from datasheet
-   send4bitI2C(0x33, false, false, false)
-   send4bitI2C(0x32, false, false, false)
+   -- init sequence from datasheet (Figure 24)
+   local function justsend(what)
+     i2c.start(busid)
+     i2c.address(busid, busad, i2c.TRANSMITTER)
+     i2c.write(busid, bit.set(what, en))
+     i2c.write(busid, what)
+     i2c.stop(busid)
+   end
+   local three = bit.bor(bit.bit(d4), bit.bit(d5))
+   justsend(three)
+   tmr.delay(5)
+   justsend(three)
+   tmr.delay(1)
+   justsend(three)
+   tmr.delay(1)
+   justsend(bit.bit(d5))
+   -- we are now primed for the FUNCTIONSET command from the liquidcrystal ctor
 
    -- Return backend object
    return {
       fourbits  = true,
       command   = function (_, cmd)
-         return send4bitI2C(cmd, false, false, false)
+         return send4bitI2C(cmd, false, false)
       end,
       busy      = function(_)
-         local rv = send4bitI2C(0xff, false, true, true)
-         send4bitI2C(bit.bor(0x80, bit.clear(rv, 7)), false, false, false)
-         return bit.isset(rv, 7)
+         return bit.isset(send4bitI2C(0xff, false, true), 7)
       end,
       position  = function(_)
-         local rv = bit.clear(send4bitI2C(0xff, false, true, true), 7)
-         send4bitI2C(bit.bor(0x80, rv), false, false, false)
```

Comment:
To be hones, I don't remember what these `send4bitI2C(...(0x80, ...` lines do. Are you sure, that they are not needed?

Reply:
I think they were trying to restore the cursor position.  As I read the data sheet, which is not the clearest thing in the world, they are wrong: if the busy flag is set, no commands should be sent.

Reply:
I have some fuzzy memories about screen not accepting any input unless given this command. Guess it's not needed since the example worked.

Reply:
I wonder just how faithful all the HD44780 implementations floating around are.  If we find one that needs special handling -- or indeed, that the version I have is itself "special" in some way -- we can revisit.

Reply:
No, I've tested this code on the same unit. Not sure that I want to dig deeper into this:)